src/Tasks/Microsoft.NET.Build.Tasks/PrepareForReadyToRunCompilation.cs

```diff
@@ -248,16 +214,131 @@ protected override void ExecuteCore()
+                        {
+                            r2rSymbolsFileToPublish.SetMetadata(MetadataKeys.ExcludeFromSingleFile, "true");
+                        }
+
+                        r2rFilesPublishList.Add(r2rSymbolsFileToPublish);
+                    }
+                }
+
+                imageCompilationList.Add(r2rCompilationEntry);
+
+                // Publish it
+                TaskItem compositeR2RFileToPublish = new TaskItem(MainAssembly);
+                compositeR2RFileToPublish.ItemSpec = compositeR2RImage;
+                compositeR2RFileToPublish.RemoveMetadata(MetadataKeys.OriginalItemSpec);
+                compositeR2RFileToPublish.SetMetadata(MetadataKeys.RelativePath, compositeR2RImageRelativePath);
+                r2rFilesPublishList.Add(compositeR2RFileToPublish);
+            }
         }
 
-        private enum Eligibility
+        private struct Eligibility
         {
-            None,
-            ReferenceOnly,
-            CompileAndReference
+            [Flags]
+            private enum EligibilityEnum
+            {
+                None = 0,
+                Reference = 1,
+                HideReferenceFromComposite = 2,
+                CompileSeparately = 4,
+                CompileIntoCompositeImage = 8,
+            }
+
+            private readonly EligibilityEnum _flags;
+
+            public static Eligibility None => new Eligibility(EligibilityEnum.None);
+
+            public bool NoEligibility => _flags == EligibilityEnum.None;
+            public bool IsReference => (_flags & EligibilityEnum.Reference) == EligibilityEnum.Reference;
+            public bool ReferenceHiddenFromCompositeBuild => (_flags & EligibilityEnum.HideReferenceFromComposite) == EligibilityEnum.HideReferenceFromComposite;
+            public bool CompileIntoCompositeImage => (_flags & EligibilityEnum.CompileIntoCompositeImage) == EligibilityEnum.CompileIntoCompositeImage;
+            public bool CompileSeparately => (_flags & EligibilityEnum.CompileSeparately) == EligibilityEnum.CompileSeparately;
+            public bool Compile => CompileIntoCompositeImage || CompileSeparately;
+
+            private Eligibility(EligibilityEnum flags)
+            {
+                _flags = flags;
+            }
+
+            public static Eligibility CreateReferenceEligibility(bool hideFromCompositeBuilds)
+            {
+                if (hideFromCompositeBuilds)
+                    return new Eligibility(EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite);
+                else
+                    return new Eligibility(EligibilityEnum.Reference);
+            }
+
+            public static Eligibility CreateCompileEligibility(bool doNotBuildIntoComposite)
+            {
+                if (doNotBuildIntoComposite)
+                    return new Eligibility(EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite | EligibilityEnum.CompileSeparately);
```

Comment:
All of this looks great, I'm just somewhat confused what is the difference between <code>EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite</code> and <code>EligibilityEnum.Reference | EligibilityEnum.HideReferenceFromComposite | EligibilityEnum.CompileSeparately</code>, can you please elaborate?

Reply:
The first will represent a dll which is referenced only by separate R2R compilation, and the second is one which is actually compiled separately into a unique dll via crossgen2 or crossgen. Effectively, no code will be included from the assembly into a compiled R2R image unless one of the `Compile` flags is set.